### PR TITLE
Tiny chores

### PR DESCRIPTION
**Pull request by mike-sul**
Refactor cmakefiles and add tests folder    
Add tests folder.
Restructure cmakefiles to make building and running tests slightly easier.
Add basic stuff into README.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -3,30 +3,44 @@
 # SPDX-License-Identifier: MIT
 
 cmake_minimum_required (VERSION 3.5)
+project(aklite)
 
-# Export compile_commands.json for clang-tidy
-set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+# OSTree and DockerApp package manager are mandatory for aktualizr-lite
+set(BUILD_OSTREE ON CACHE BOOL "")
+set(BUILD_DOCKERAPP ON CACHE BOOL "")
```
Comment (doanac) [inline CMakeLists.txt:10]: DockerApp isn't required. We actually have at least one platform without docker support (riscv).

File: cmake-init.sh [hunk 2]
```diff
@@ -9,4 +9,4 @@ if [ "$HERE" = "$(pwd)" ] ; then
 	exit 1
 fi
 
-cmake -GNinja -DBUILD_OSTREE=ON -DBUILD_DOCKERAPP=ON $HERE
+cmake -GNinja $HERE
```
Comment (doanac) [inline cmake-init.sh:12]: you'll probably want/need -DBUILD_DOCKERAPP=ON back

File: unit-test [hunk 3]
```diff
@@ -14,9 +14,10 @@ build=$(mktemp -d -p $here)
 cd $build
 
 ../cmake-init.sh
-ninja aktualizr-lite t_lite-helpers libt_lite-mock.so uptane-generator make_ostree_sysroot
+ninja aklite-tests -j6
```
Comment (doanac) [inline unit-test:17]: ninja can't automatically determine this?

**Comment by doanac** [comment]
I like the general idea of this. I know next to nothing on CMake and this is moving things to look a lot cleaner.

File: CMakeLists.txt (hunk 1 above)
Comment (mike-sul) [reply CMakeLists.txt:10]: OK, will make BUILD_DOCKERAPP  back as optional.

File: unit-test (hunk 3 above)
Comment (mike-sul) [reply unit-test:17]: apparently yes, just get used to `make` and never realized that ninja can do it for you :)

File: README.adoc [hunk 4]
```diff
@@ -0,0 +1,69 @@
+[discrete]
+= aktualizr-lite
```
Comment (doanac) [inline README.adoc:2]: You think doing this as Markdown would make sense? That's pretty much the defacto standard of most projects these days. and we aren't building docs from this like the aktualizr repo does?

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -1,28 +1,32 @@
...
-target_link_libraries(aktualizr-lite aktualizr_lib)
+add_executable(${TARGET} ${SRC})
 
-install(TARGETS aktualizr-lite RUNTIME DESTINATION bin COMPONENT aktualizr-lite)
+if(CLANG_TIDY)
+set_target_properties(${TARGET}
+  PROPERTIES CXX_CLANG_TIDY
+# TODO: consider adding clang-tidy options that it makes sense to include here
+  clang-tidy-6.0;-checks=*;-warnings-as-errors=*
+)
```
Comment (doanac) [inline src/CMakeLists.txt:12]: I assume this is the same list that aktualizr is using?

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  ${AKLITE_DIR}/src
+  ${AKTUALIZR_DIR}/src/libaktualizr
+  ${AKTUALIZR_DIR}/third_party/jsoncpp/include
+  ${GLIB_INCLUDE_DIRS}
+  ${LIBOSTREE_INCLUDE_DIRS}
+)
+target_link_libraries(t_lite-helpers aktualizr_lib ${Boost_LIBRARIES})
+
+# TODO: Find a way to test without a need in the lite-mock
+set_tests_properties(test_lite-helpers PROPERTIES ENVIRONMENT LD_PRELOAD=$<TARGET_FILE:lite-mock>)
```
Comment (doanac) [inline tests/CMakeLists.txt:39]: We could do something like: https://github.com/advancedtelematic/aktualizr/blob/master/src/libaktualizr/package_manager/dockerappmanager_test.cc#L21-L39

File: unit-test [hunk 7]
```diff
@@ -13,10 +13,21 @@ set -x
...
+../cmake-init.sh -DCLANG_TIDY=ON
+ninja aklite-tests
 
 ctest -V -R test_lite-helpers
 ctest -V -R test_aktualizr-lite
+
+# test building the sota tools
+rm -rf ./*
+../cmake-init.sh -DBUILD_DOCKERAPP=OFF -DBUILD_SOTA_TOOLS=ON
+ninja all
```
Comment (doanac) [inline unit-test:25]: does this mean we'll be doing clang tidy against all of aktualizr also? or just the stuff under `src` and `tests`?

File: README.adoc (hunk 4 above)
Comment (mike-sul) [reply README.adoc:2]: > You think doing this as Markdown would make sense?
I don't have a strong opinion. There are different opinions around, https://mister-gold.pro/posts/en/asciidoc-vs-markdown/, https://www.zverovich.net/2016/06/16/rst-vs-markdown.html. I have worked mostly with asciidoc and RST. I suppose if we are not going to create huge documentation then Markdown should suffice.

File: src/CMakeLists.txt (hunk 5 above)
Comment (mike-sul) [reply src/CMakeLists.txt:12]: Not exactly, that's what aktualizr does https://github.com/advancedtelematic/aktualizr/blob/9b0ce2144c8d4630aa7ee3fcca2c603df493bb56/CMakeLists.txt#L278. I don't think that the diff is critical.
The given approach cleaner as it makes use of the make built-in functionality while the aktualizr implements their own functionality and introduces a custom target for each source file. In terms of clang-tidy options, we can put there whatever we consider it makes sense to do.

File: tests/CMakeLists.txt (hunk 6 above)
Comment (mike-sul) [reply tests/CMakeLists.txt:39]: That trick works for C/C++ based tests, but it won't work for high-level tests written in bash, python, etc (e.g. test_lite.sh). I figured out the way how to get rid both the mock library and the trick implemented here https://github.com/advancedtelematic/aktualizr/blob/master/src/libaktualizr/package_manager/dockerappmanager_test.cc#L21-L39.

It requires changes either here https://github.com/advancedtelematic/aktualizr/blob/9b0ce2144c8d4630aa7ee3fcca2c603df493bb56/src/libaktualizr/package_manager/ostreemanager.cc#L324 or overriding OstreeManager::getCurrent() in DockerAppManager if Patrick wouldn't like to accept that change.

File: unit-test (hunk 7 above)
Comment (mike-sul) [reply unit-test:25]: Just stuff under src as it was before, I can make it for both src and tests, but currently, clang-tody is not happy about the stuff under tests, which implies that I have to make it happy before apply it.

Comment (doanac) [reply unit-test:25]: oops - sorry about that. don't worry about it on tests for now.

File: tests/CMakeLists.txt (hunk 6 above)
Comment (mike-sul) [reply tests/CMakeLists.txt:39]: In any case, I won't do it in this PR.

**Comment by mike-sul** [comment]
@doanac Is this PR good enough for merging?

**Comment by doanac** [comment]
i'll give it a shot today. Looks good though.

**Comment by doanac** [comment]
I like the general look. But I don't think the ctest and clang are actually working now:

https://ci.foundries.io/projects/aklite/builds/47/unit-test/artifacts/console.log
~~~
+ ctest -V -R test_lite-helpers
UpdateCTestConfiguration  from :/repo/tmp.7RoR0Swq37/DartConfiguration.tcl
Parse Config file:/repo/tmp.7RoR0Swq37/DartConfiguration.tcl
UpdateCTestConfiguration  from :/repo/tmp.7RoR0Swq37/DartConfiguration.tcl
Parse Config file:/repo/tmp.7RoR0Swq37/DartConfiguration.tcl
Test project /repo/tmp.7RoR0Swq37
Constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
No tests were found!!!
+ ctest -V -R test_aktualizr-lite
UpdateCTestConfiguration  from :/repo/tmp.7RoR0Swq37/DartConfiguration.tcl
Parse Config file:/repo/tmp.7RoR0Swq37/DartConfiguration.tcl
UpdateCTestConfiguration  from :/repo/tmp.7RoR0Swq37/DartConfiguration.tcl
Parse Config file:/repo/tmp.7RoR0Swq37/DartConfiguration.tcl
Test project /repo/tmp.7RoR0Swq37
Constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
No tests were found!!!
~~~

I guess in addition to fixing that. Could you also share how, when its working, that I could run clang against these files. eg what's the new way to do `ninja aktualizr_clang_tidy-src-helpers.cc  aktualizr_clang_tidy-src-main.cc`

**Comment by mike-sul** [comment]
> I like the general look. But I don't think the ctest and clang are actually working now:

Yeah, I screwed it up, shouldn't have touched something that worked :).

**Comment by mike-sul** [comment]
I fixed those ctest and clan-tidy issues.
Regarding ctest it was just stupid mistake - `include(Test)` (or enable_testing) has to be in the root make file.

Regarding `tidy`. Actually, it was a stupid mistake too, I just didn't put braces around the clang-tidy cmd call in the target property definition. The given approach works per-target basis and run clang-tidy during compilation. I like this approach because

1)  you don't have to produce and maintain a bunch of custom functions - just make use of the make builtin feature;

2) you don't generate huge amount of custom/additional targets - one target per each source file (`cmake --build . --target help | wc -l` yields 844);

3) you don't have to remember/find a "clang-tidy" custom target name of each source, instead you just need to know about your regular targets (e.g. aktualizr-lite) and calling ninja|make aktualizr-lite will do "tidying";

Drawbacks of this approach that you have to turn on/off cmake configuration variable (`-DBUILD_WITH_CLANG_TIDY` and applies to all sources. So you cannot compile your sources without clang-tidy and them just run clang-tidy for a desired source file(s) without build reconfiguration.

With that said, I just enabled the both approaches, the aktualizr's and the one with defining `CXX_CLANG_TIDY` property of a target (or the whole project). Therefore, we can use one of them of both depending on what is more convenient at a given situation.

**Comment by mike-sul** [comment]
This is how it looks when clang-tidy is part of a compilation process.

`[ 97%] Building CXX object src/CMakeFiles/aktualizr-lite.dir/main.cc.o
cd /home/mike/work/foundries/projects/aklite/build/src && /usr/bin/ccache /usr/local/bin/cmake -E __run_co_compile --tidy="/usr/bin/clang-tidy-6.0;-quiet;-header-filter=(/home/mike/work/foundries/projects/aklite|\\.\\.)/src/.*" --source=/home/mike/work/foundries/projects/aklite/src/main.cc -- /usr/bin/c++  -DBOOST_ALL_NO_LIB -DBOOST_ATOMIC_DYN_LINK -DBOOST_CHRONO_DYN_LINK -DBOOST_DATE_TIME_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_LOG_DYN_LINK -DBOOST_LOG_SETUP_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DBUILD_DOCKERAPP -I/home/mike/work/foundries/projects/aklite/aktualizr/src/libaktualizr -I/home/mike/work/foundries/projects/aklite/aktualizr/third_party/jsoncpp/include -I/home/mike/work/foundries/projects/aklite/aktualizr/third_party/googletest/googletest/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/ostree-1  -g   -o CMakeFiles/aktualizr-lite.dir/main.cc.o -c /home/mike/work/foundries/projects/aklite/src/main.cc
`

**Comment by doanac** [comment]
Its working for me now. I think we are ready to merge. Did you want to try and clean up the commit history, or just merge with a merge-commit?

**Comment by mike-sul** [comment]
> Its working for me now. I think we are ready to merge. Did you want to try and clean up the commit history, or just merge with a merge-commit?

I don't have a strong opinion on it, the button "Rebase and merge" that we agreed on the slack channel earlier works for me, although we can "squash and merge" or I can squash locally into a single commit and then press "Rebase and merge". It's up to you :).

**Comment by doanac** [comment]
I was going to say rebase and merge. However that might not be bisectable.
Let's squash and merge

**Pull request merged**